lib/foreman_maintain/package_manager/yum.rb

```diff
@@ -0,0 +1,108 @@
+
+    VERSIONLOCK_START_CLAUSE = '## foreman-maintain - start'
+    VERSIONLOCK_END_CLAUSE = '## foreman-maintain - end'
+    VERSIONLOCK_CONFIG_FILE = '/etc/yum/pluginconf.d/versionlock.conf'
+
+    def foreman_related_packages
+      installed = sys.execute("repoquery -a --qf='%{envra} %{repo.baseurl}'")
+                    .split("\n"). map(&:split)
+      repos_to_lock = []
+      foreman_repo = installed.find { |package| package[0] =~ /foreman-installer-\d+/ }[1]
+      repos_to_lock << foreman_repo
+      installed.inject([]) do |related, package|
+        if repos_to_lock.include?(package[1])
+          pkg_info = parse_envra(package[0])
+          related << pkg_info if pkg_info
+        end
+        related
+      end
+    end
+
+    def version_locking_packages
+      %w[yum-utils yum-plugin-versionlock]
+    end
+
+    def lock_versions(package_list)
+      unlock_versions
+      open(versionlock_file, 'a') do |f|
+        f.puts VERSIONLOCK_START_CLAUSE
+        f.puts '# The following packages are locked by foreman-maintain. Do not modify!'
+        package_list.each { |package| f.puts "#{package[:envra]}.*" }
+        f. puts '# End of list of packages locked by foreman-maintain'
+        f.puts VERSIONLOCK_END_CLAUSE
+      end
+    end
+
+    def unlock_versions
+      lock_file = versionlock_file
+      content = File.read(lock_file)
+      content = content.gsub(/#{VERSIONLOCK_START_CLAUSE}.*#{VERSIONLOCK_END_CLAUSE}\n/m, '')
+      File.open(lock_file, 'w') { |f| f.write content }
+    end
+
+    def version_locking_enabled?
+      installed?(version_locking_packages) && versionlock_config =~ /^\s*enabled\s+=\s+1/
+    end
+
+    # make sure the version locking tools are configured
+    #  enabled = 1
+    #  locklist = <list file>
+    # we can assume it is already installed
+    def configure_version_locking
+      config = versionlock_config
+      config += "\n" unless config[-1] == "\n"
+      enabled_re = /^\s*enabled\s*=.*$/
+      if enabled_re.match(config)
+        config = config.gsub(enabled_re, 'enabled = 1')
+      else
+        config += "enabled = 1\n"
+      end
+      unless config =~ /^\s*locklist\s*=.*$/
+        config += "locklist = /etc/yum/pluginconf.d/versionlock.list\n"
+      end
+      File.open(VERSIONLOCK_CONFIG_FILE, "w") {|file| file.puts config }
+    end
+
+    def installed?(packages)
+      packages_list = [packages].flatten(1).map { |pkg| "'#{pkg}'" }.join(' ')
+      sys.execute?(%(rpm -q #{packages_list}))
+    end
+
+    def install(packages, assumeyes: false)
+      yum_action('install', packages, assumeyes: assumeyes)
+    end
+
+    def update(packages, assumeyes: false)
+      yum_action('update', packages, assumeyes: assumeyes)
+    end
+
+    private
+
+    def versionlock_config
+      File.read(VERSIONLOCK_CONFIG_FILE) if File.exist?(VERSIONLOCK_CONFIG_FILE)
+    end
+
+    def versionlock_file
+      /^\s*locklist\s*=\s*(\S+)/.match(versionlock_config).captures[0]
```

Comment:
In case of `versionlock_config_file` not present then Nil case handling required.

Reply:
@kgaikwad good point. I made it a bit more robust.